proposals/2023/5-11_queue-sensitivity.md

```diff
@@ -0,0 +1,188 @@
+
+
+**Please briefly answer the following questions:**
+
+1. Why is this required?
+
+<!-- Answer here -->
+
+* improves security by removing potentially sensitive pipeline information from the queue
+* takes a greater "dumb worker" approach by giving them less direct-access to potentially sensitive data
+* moves access to sensitive pipeline information behind the worker auth & build token process
+
+1. If this is a redesign or refactor, what issues exist in the current implementation?
+
+<!-- Answer here -->
+
+The current queue implementation involves placing sensitive pipeline information like the user token in Redis, and the worker depends on this information being present to properly execute builds.
+
+
+1. Are there any other workarounds, and if so, what are the drawbacks?
+
+<!-- Answer here -->
+
+It's possible to improve security around the queue in other ways, like using a [READONLY](https://redis.io/commands/readonly/) replica, though it doesn't improve the sensitivity of the data that can be read.
+
+We could also look into GitHub Apps to limit the blast radius if the queue were compromised as an attack vector.
+
+1. Are there any related issues? Please provide them below if any exist.
+
+<!-- Answer here -->
+
+Indirect: https://github.com/go-vela/community/issues/804
+
+## Design
+
+<!--
+This section is intended to explain the solution design for the proposal.
+
+NOTE: If there are no current plans for a solution, please leave this section blank.
+-->
+
+**Please describe your solution to the proposal. This includes, but is not limited to:**
+
+* new database table to store interim compiled pipelines
+* new endpoint to retrieve a compiled pipeline from the server
+* modification to the [Item](https://github.com/go-vela/types/blob/main/item.go#L12C1-L18) currently placed on the queue to only push build ID
+* modification to webhook processing to push the compiled pipeline to a separate table
+* added interaction between worker and server to request a compiled pipeline
+
+<!-- Answer here -->
+
+For reference please review the following design diagram that explains the approach from a high level.
+
+```mermaid
+flowchart TD
+    A[GitHub] --> B[Vela Server]
+    B --> |compile\npush compiled| C[Redis Queue]
+    C --> |pop compiled| D[Vela Worker]
+    D --> E(execute build)
+
+    F[GitHub] --> G[Vela Server]
+    G --> |compile\npush compiled| H[Database]
+    G --> |push build_id| K[Redis Queue]
+    K --> |pop build_id| I[Vela Worker]
+    I --> |request compiled from server\nvia build_token| G
+    I --> J(execute build)
+```
+
+New functionality that is absolutely required:
+
+#### Server
+
+- GET `api/v1/<org>/<repo>/build/<number>/compiled` or `api/v1/search/builds/<id>/compiled` to retrieve a compiled pipeline using a build token
+- `database/compiled` package and a new database table `compiled` that will contain precompiled pipelines in bytes, formerly stored directly on the Redis queue.
```

Comment:
I guess most of my confusion comes from the `PostWebhook` method which has `p`, `b`, `pipeline`, and `compiled` vars. Keeping track of what's what and why the vars are different is confusing. esp because `compiled` gets saved to the database as a `pipeline`.

https://github.com/go-vela/server/blob/main/api/webhook.go#L133
https://github.com/go-vela/server/blob/main/api/webhook.go#L410-L413
https://github.com/go-vela/server/blob/main/api/webhook.go#L434
https://github.com/go-vela/server/blob/main/api/webhook.go#L508-L518
https://github.com/go-vela/server/blob/main/api/webhook.go#L561-L562
https://github.com/go-vela/server/blob/main/api/webhook.go#L697-L705
```go
func PostWebhook(c *gin.Context) {
	...
	h, r, b := webhook.Hook, webhook.Repo, webhook.Build // *library.Hook, *library.Repo, *library.Build
	...
	var(
		// variable to store the raw pipeline configuration
		config []byte
		// variable to store executable pipeline
		p *pipeline.Build
		// variable to store pipeline configuration
		pipeline *library.Pipeline
	)
	...
	pipeline, err = database.FromContext(c).GetPipelineForRepo(b.GetCommit(), repo)
	if err != nil { // assume the pipeline doesn't exist in the database yet
		// send API call to capture the pipeline configuration file
		config, err = scm.FromContext(c).ConfigBackoff(u, repo, b.GetCommit())
		...
	} else {
		config = pipeline.GetData()
	}
	...
	var compiled *library.Pipeline
	// parse and compile the pipeline configuration file
	p, compiled, err = compiler.FromContext(c).
		Duplicate().
		WithBuild(b).
		WithComment(webhook.Comment).
		WithFiles(files).
		WithMetadata(m).
		WithRepo(repo).
		WithUser(u).
		Compile(config)
	...
	if pipeline == nil {
		pipeline = compiled
		err = database.FromContext(c).CreatePipeline(pipeline)
	...
	// publish the build to the queue
	go publishToQueue(
		queue.FromGinContext(c),
		database.FromContext(c),
		p,
		b,
		repo,
		u,
	)
)
...
func publishToQueue(queue queue.Service, db database.Service, p *pipeline.Build, b *library.Build, r *library.Repo, u *library.User) {
	item := types.ToItem(p, b, r, u)
...
```

So, I guess these are related like this, right?
- `database/build` -> `library/build` event-based metadata for a build
- `database/pipeline` -> `library/pipeline` the raw pipeline for a SHA
- `database/compiled` -> `pipeline/build` the compiled pipeline = raw SHA pipeline + build metadata

If that's right, then maybe the `types/pipeline` module is effectively a `library/compiled` module which means `pipeline/build` is logically a `library/compiled/build`?

When implementing this proposal, I hope that at least the `PostWebhook` method becomes clearer and easier to follow.

Reply:
yeah thats great feedback thanks @cognifloyd 
i have the same understanding of the 3 builds tables but i agree theres room to improve the clarity on these. 
i see it like this:
`build` -> as event-based metadata that is generated through processing the webhook
`pipeline` -> as the raw vela pipeline data that is attached to a commit in git
`compiled` -> as the executable pipeline in JSON that is derived from the pipeline+webhook-data, which is different from pipeline as it contains the substituted metadata and anything purged from, say, rulesets for example.

to be honest im not completely in love with the name `compiled` because its a bit vague and only implies that the pipeline YAML has been compiled by the server... which is true but its not completely accurate.

Reply:
> Is this right?
> flowchart LR
>     A[Pipeline] --> |1:n| B[Build]
>     B --> |1:1| C[Compiled]

Yeah, exactly. Naming is tricky when all three things are somewhat all related... 

The way I like to think about it is:
`library.Build` -> Vela's record of the build (metadata mostly)
`library.Pipeline` -> What the contents of the `.vela.yml` are at any given commit
`pipeline.Build` -> The script we give to the worker to run the build

I'm glad you pointed out this line here:
```go
	p, compiled, err = compiler.FromContext(c).
		Duplicate().
		WithBuild(b).
		WithComment(webhook.Comment).
		WithFiles(files).
		WithMetadata(m).
		WithRepo(repo).
		WithUser(u).
		Compile(config)
```

Perhaps there is a better name than `compiled`. Because in this case the `p` is a `pipeline.Build` whereas the `compiled` is a `library.Pipeline`.

Reply:
Maybe instead of `Compiled` we could use `ExecutablePipeline`?

Reply:
I like `executable`!

Reply:
I think `ExecutableBuild` would be even better, because it is a particular build that the worker runs.

So my chart ends up looking like this:
```mermaid
flowchart LR
    A[Pipeline] --> |1:n| B[Build]
    B --> |1:1| C[ExecutableBuild]
```

Reply:
```suggestion
- GET `api/v1/<org>/<repo>/build/<number>/executable` or `api/v1/search/builds/<id>/executable` to retrieve a compiled pipeline using a build token
- `database/executable` package and a new database table `executable` that will contain executable pipeline builds in bytes, formerly stored directly on the Redis queue.
```

Reply:
i like executable build, also